Guardian: Introduce additional constructors

papagian opened the pull request:
<!--

Thank you for sending a pull request! Here are some tips:

1. If this is your first time, please read our contribution guide at https://github.com/grafana/grafana/blob/main/CONTRIBUTING.md

2. Ensure you include and run the appropriate tests as part of your Pull Request.

3. In a new feature or configuration option, an update to the documentation is necessary. Everything related to the documentation is under the docs folder in the root of the repository.

4. If the Pull Request is a work in progress, make use of GitHub's "Draft PR" feature and mark it as such.

5. If you can not merge your Pull Request due to a merge conflict, Rebase it. This gets it in sync with the main branch.

6. Name your PR as "<FeatureArea>: Describe your change", e.g. Alerting: Prevent race condition. If it's a fix or feature relevant for the changelog describe the user impact in the title. The PR title is used to auto-generate the changelog for issues marked with the "add to changelog" label.

7. If your PR content should be added to the What's New document for the next major or minor release, add the **add to what's new** label to your PR. Note that you should add this label to the main PR that introduces the feature; do not add this label to smaller PRs for the feature.

-->

**What is this feature?**
This fix introduces two more dashboard guardian constructors (in addition to the existing one that expects a dashboard ID as input):
- one expecting the dashboard UID and
- another expecting the dashboard (`*models.Dashboard`).

The latter should be preferred  by the callers (if the dashboard is available) over the others because it avoids querying the database for fetching the dashboard.

Furthermore, updates the guardian callers to use the most appropriate constructor each time.

**Why do we need this feature?**

This change is driven by a failure to GET existing **nested folder** when checking whether the user has adequate permissions. During folder creation:

- if the nested folders are enabled the folder is inserted in both stores (for the development/beta period): the legacy (`dashboard` table) and the new (`folder` table) with [the same UID](https://github.com/grafana/grafana/blob/bf49c20050828900d7938b5055922fed19a39039/pkg/services/folder/folderimpl/folder.go#L290).
- the necessary permissions are created too:
https://github.com/grafana/grafana/blob/64143ea7d0f6d7a85e06147ee29e2fb063873049/pkg/services/folder/folderimpl/folder.go#L262-L282

However, this permissions are associated [with the folder ID](https://github.com/grafana/grafana/blob/64143ea7d0f6d7a85e06147ee29e2fb063873049/pkg/models/dashboard_acl.go#L42).
As a result when trying to access a nested folder (from the `folder` table) for checking the permissions we get a failure because the `folder` ID does not exist.


**Who is this feature for?**

Is for supporting nested folders but we believe that is aligned with the general goal of dropping dependency on the database sequential IDs.

**Which issue(s) does this PR fix?**:

<!--

- Automatically closes linked issue when the Pull Request is merged.

Usage: "Fixes #<issue number>", or "Fixes (paste link of issue)"

-->

Fixes #

**Special notes for your reviewer**:
Since we frequently need to fetch the folder `UID` from the `ID` and the opposite we can consider as a follow-up to use some caching to avoid the additional database overhead.

papagian committed:
Guardian: Use dashboard UID instead of ID

papagian committed:
Merge remote-tracking branch 'origin/main' into papagian/dashboard-guardian-uid

papagian committed:
Fix fetching version list failure

kalleep commented on pkg/services/guardian/accesscontrol_guardian.go:
```diff
@@ -22,23 +22,36 @@ var permissionMap = map[string]models.PermissionType{
-) *AccessControlDashboardGuardian {
+) (*AccessControlDashboardGuardian, error) {
+	dashID := int64(0)
+	if dashboardUID != "" {
+		q := &models.GetDashboardQuery{
+			Uid:   dashboardUID,
+			OrgId: user.OrgID,
+		}
+
+		if err := dashboardService.GetDashboard(ctx, q); err != nil {
+			return nil, ErrGuardianGetDashboardFailure.Errorf("failed to get dashboard by UID: %w", err)
+		}
+		dashID = q.Result.Id
```
If we now fetch the dashboard in the constructor we should store the dashboard result and remove loadDashboard() function inside the guardian

kalleep commented on pkg/services/guardian/accesscontrol_guardian.go:
```diff
@@ -22,23 +22,36 @@ var permissionMap = map[string]models.PermissionType{
 		ctx:                         ctx,
 		log:                         log.New("dashboard.permissions"),
-		dashboardID:                 dashboardId,
+		dashboardID:                 dashID,
```
no need to keep the dashboardID if we remove the loadDashboard functoin

kalleep commented on pkg/api/alerting.go:
```diff
@@ -693,7 +693,15 @@ func (hs *HTTPServer) PauseAlert(legacyAlertingEnabled *bool) func(c *models.Req
 		}
 
-		guardian := guardian.New(c.Req.Context(), query.Result.DashboardId, c.OrgID, c.SignedInUser)
+		dashUID, err := hs.getDashboardUID(c.Req.Context(), query.Result.DashboardId, c.OrgID)
+		if err != nil {
+			return response.Error(500, "Get dashboard failed", err)
+		}
+
+		guardian, err := guardian.New(c.Req.Context(), dashUID, c.OrgID, c.SignedInUser)
+		if err != nil {
+			return response.ErrOrFallback(http.StatusInternalServerError, "Error while creating permission guardian", err)
+		}
```
As is this will add additional db queries to check permissions on a dashboard. First translate id to uid, then in guardian trasnalte the uid to id. Seems like we should add additional constructors for dashboard guardian e.g. NewFromID(), NewFromUID() and maybe one for when we aleardy have the full dashboard model available (to skip additional query to get model)

kalleep reviewed (commented):
Nice, left some comments on things I think we can improve

papagian committed:
Apply suggestions from code review

Introduce several guardian constructors and each time use
the most appropriate one.

papagian committed:
fixup

papagian commented:
> Nice, left some comments on things I think we can improve

I have made the suggested changes; can you check once again, thank you!

papagian committed:
Merge remote-tracking branch 'origin/main' into papagian/dashboard-guardian-uid

papagian committed:
Merge remote-tracking branch 'origin/main' into papagian/dashboard-guardian-uid

kalleep reviewed (approved):
LGTM! Nice work
